### PR TITLE
[UNR-1009] Use events to send Unreliable RPCs

### PR DESCRIPTION
**Pull request by mattyoung-improbable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Using commands for Unreliable RPCs is causing performance issues. This PR changes them to use eventss while keeping Commands for Reliable RPCs.

#### Release note
Feature: Unreliable RPCs now implemented using events instead of commands.

#### Tests
Tested locally, verified reliable / unreliable client/server/multicast rpc types still working, only called on relevant clients.

QA: Bugs to look for include movement issues and other player actions not working correctly. Also game wide events / SFX / Particle FX.

### CODE REVIEW

File: SpatialGDK/Extras/schema/rpc_components.schema [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+package unreal;
+
+import "unreal/gdk/core_types.schema";
```
Comment (Vatyx) [inline SpatialGDK/Extras/schema/rpc_components.schema]: Since this is the only file using `UnrealRPCCommandResponse` now, I wanted to move that type into this file so we don't have to include core_types.schema

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp [hunk 2]
```diff
@@ -40,46 +40,64 @@ FORCEINLINE UClass* ResolveClass(FString& ClassPath)
...
 	{
-		ESchemaComponentType RPCType = SCHEMA_Invalid;
-		if (RemoteFunction->FunctionFlags & FUNC_NetClient)
-		{
-			RPCType = SCHEMA_ClientRPC;
-		}
-		else if (RemoteFunction->FunctionFlags & FUNC_NetServer)
+		return SCHEMA_NetMulticastRPC;
+	}
+	if (Flags & FUNC_NetCrossServer)
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp]: Can we use `HasAnyFunctionFlags` for all of these ifs? Could we also make the ifs into else ifs?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -840,17 +848,70 @@ void USpatialReceiver::OnComponentUpdate(Worker_ComponentUpdateOp& Op)
 	}
 	else if (Category == ESchemaComponentType::SCHEMA_NetMulticastRPC)
 	{
-		if (const TArray<UFunction*>* RPCArray = Info.RPCs.Find(SCHEMA_NetMulticastRPC))
-		{
-			ReceiveMulticastUpdate(Op.update, TargetObject, *RPCArray);
-		}
+		checkNoEntry();
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: I don't think this else if has to exist anymore.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp [hunk 4]
```diff
@@ -40,46 +40,64 @@ FORCEINLINE UClass* ResolveClass(FString& ClassPath)
...
+	for (UFunction* RemoteFunction : RelevantClassFunctions)
+	{
+		ESchemaComponentType RPCType = GetRpcType(RemoteFunction);
+		checkf(RPCType != SCHEMA_Invalid, TEXT("Could not determine RPCType for RemoteFunction: %s"), *GetPathNameSafe(RemoteFunction));
+		
 		FRPCInfo RPCInfo;
 		RPCInfo.Type = RPCType;
-		RPCInfo.Index = RPCArray.Num();
+		RPCInfo.bReliable = RemoteFunction->HasAnyFunctionFlags(FUNC_NetReliable);
+		RPCInfo.Index = Info->RPCs.Num();
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp]: Can you put a comment here regarding how we can uniquely identify RPCs based on this index on both clients and servers

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -840,17 +848,70 @@ void USpatialReceiver::OnComponentUpdate(Worker_ComponentUpdateOp& Op)
...
+				return;
+			}
+		}
+		break;
+		default:
+		break;
+	}
+	Schema_Object* EventsObject = Schema_GetComponentUpdateEvents(Op.update.schema_type);
+
+	uint32 EventCount = Schema_GetObjectCount(EventsObject, 1);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Like we do in the GlobalStateManager, could we make constants for these field ids rather than having a naked 1.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 6]
```diff
@@ -117,8 +120,7 @@ Worker_RequestId USpatialSender::CreateEntity(USpatialActorChannel* Channel)
 			return;
 		}
 
-		WorkerRequirementSet& RequirementSet = Type == SCHEMA_ClientRPC ? OwningClientOnly : ServersOnly;
-		ComponentWriteAcl.Add(ComponentId, RequirementSet);
+		ComponentWriteAcl.Add(ComponentId, ServersOnly);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Why this change? Think I'm missing something basic.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 7]
```diff
@@ -443,8 +442,24 @@ void USpatialSender::SendRPC(TSharedRef<FPendingRPCParams> Params)
...
+		else if (RPCInfo->Type == SCHEMA_ServerUnreliableRPC)
+		{
+			ComponentId = SpatialConstants::CLIENT_RPCS_COMPONENT_ID;
+		}
+		else
+		{
+			ComponentId = SpatialConstants::NETMULTICAST_RPCS_COMPONENT_ID;
+		}
+
+		Worker_ComponentUpdate ComponentUpdate = CreateMulticastUpdate(TargetObject, Params->Function, Params->Parameters.GetData(), ComponentId, RPCInfo->Index, EntityId, UnresolvedObject);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Can `CreateMulticastUpdate` be renamed to be more representative.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 8]
```diff
@@ -616,7 +631,7 @@ Worker_CommandRequest USpatialSender::CreateRPCCommandRequest(UObject* TargetObj
 {
 	Worker_CommandRequest CommandRequest = {};
 	CommandRequest.component_id = ComponentId;
-	CommandRequest.schema_type = Schema_CreateCommandRequest(ComponentId, CommandIndex);
+	CommandRequest.schema_type = Schema_CreateCommandRequest(ComponentId, 1);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Same here, lets put a constant instead of a magic number.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 9]
```diff
@@ -694,7 +711,9 @@ Worker_ComponentUpdate USpatialSender::CreateMulticastUpdate(UObject* TargetObje
 		}
 	}
 
-	AddBytesToSchema(EventData, 1, PayloadWriter);
+	Schema_AddUint32(EventData, 1, TargetObjectRef.Offset);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: This code is duplicated. Can we make a function that takes in a Schema_Object and writes an RPC payload?

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialClassInfoManager.h [hunk 10]
```diff
@@ -33,6 +33,7 @@ struct FRPCInfo
 {
 	ESchemaComponentType Type;
 	uint32 Index;
+	bool bReliable;
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialClassInfoManager.h]: I think this can be deleted.

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h [hunk 11]
```diff
@@ -23,13 +23,16 @@ enum ESchemaComponentType : int32
 	SCHEMA_OwnerOnly,
 	SCHEMA_Handover,
 
+	SCHEMA_Count,
+
 	// RPCs
 	SCHEMA_ClientRPC,
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h]: Would you want to change this to be `SCHEMA_ClientReliableRPC` (Ditto for server)

File: SpatialGDK/Extras/schema/rpc_components.schema (hunk 1 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Extras/schema/rpc_components.schema]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp (hunk 2 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp]: Done

File: SpatialGDK/Extras/schema/core_types.schema [hunk 12]
```diff
@@ -6,11 +6,4 @@ type UnrealObjectRef {
...
     option<UnrealObjectRef> outer = 4;
-}
-
-type UnrealRPCCommandRequest {
-	bytes rpc_payload = 1;
-}
-
-type UnrealRPCCommandResponse {
-}
+}
```
Comment (Vatyx) [inline SpatialGDK/Extras/schema/core_types.schema]: Add new line here.

File: SpatialGDK/Extras/schema/rpc_components.schema [hunk 13]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+package unreal;
+
+type UnrealRPCPayload {
+    uint32 offset = 1;
+    uint32 rpc_index = 2;
+    bytes rpc_payload = 3;
+}
+
+type Void {}
```
Comment (Vatyx) [inline SpatialGDK/Extras/schema/rpc_components.schema]: Nit: can you make the braces match the other empty types we have: the closing brace is on the second line.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialClassInfoManager.h (hunk 10 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialClassInfoManager.h]: Done

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h (hunk 11 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 14]
```diff
@@ -653,19 +662,18 @@ Worker_CommandRequest USpatialSender::CreateRPCCommandRequest(UObject* TargetObj
...
-Worker_ComponentUpdate USpatialSender::CreateMulticastUpdate(UObject* TargetObject, UFunction* Function, void* Parameters, Worker_ComponentId ComponentId, Schema_FieldId EventIndex, Worker_EntityId& OutEntityId, const UObject*& OutUnresolvedObject)
+Worker_ComponentUpdate USpatialSender::CreateUnreliableRPCUpdate(UObject* TargetObject, UFunction* Function, void* Parameters, Worker_ComponentId ComponentId, Schema_FieldId EventIndex, Worker_EntityId& OutEntityId, const UObject*& OutUnresolvedObject)
 {
 	Worker_ComponentUpdate ComponentUpdate = {};
 
 	ComponentUpdate.component_id = ComponentId;
 	ComponentUpdate.schema_type = Schema_CreateComponentUpdate(ComponentId);
 	Schema_Object* EventsObject = Schema_GetComponentUpdateEvents(ComponentUpdate.schema_type);
-	Schema_Object* EventData = Schema_AddObject(EventsObject, EventIndex);
+	Schema_Object* EventData = Schema_AddObject(EventsObject, 1);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Nit: magic number here

File: SpatialGDK/Extras/schema/core_types.schema (hunk 12 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Extras/schema/core_types.schema]: Done

File: SpatialGDK/Extras/schema/rpc_components.schema (hunk 13 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Extras/schema/rpc_components.schema]: Done

File: SpatialGDK/Extras/schema/rpc_components.schema [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+package unreal;
+
+type UnrealRPCPayload {
+    uint32 offset = 1;
+    uint32 rpc_index = 2;
+    bytes rpc_payload = 3;
+}
+
+type Void {
+
```
Comment (Vatyx) [inline SpatialGDK/Extras/schema/rpc_components.schema]:
```suggestion
```

File: SpatialGDK/Extras/schema/rpc_components.schema [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    bytes rpc_payload = 3;
+}
+
+type Void {
+}
+
+component UnrealClientRPCEndpoint {
+    id = 9990;
+    event UnrealRPCPayload unreliable_client_to_server_rpc;
+    command Void reliable_server_to_client_rpc(UnrealRPCPayload);
```
Comment (m-samiec) [inline SpatialGDK/Extras/schema/rpc_components.schema]: Void with a capital? I think I'm going to be sick.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 17]
```diff
@@ -1441,7 +1462,7 @@ void USpatialReceiver::ReceiveRPCCommandRequest(const Worker_CommandRequest& Com
 {
 	Schema_Object* RequestObject = Schema_GetCommandRequestObject(CommandRequest.schema_type);
 
-	TArray<uint8> PayloadData = GetBytesFromSchema(RequestObject, 1);
+	TArray<uint8> PayloadData = GetBytesFromSchema(RequestObject, 3);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
	TArray<uint8> PayloadData = GetBytesFromSchema(RequestObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_PAYLOAD_ID);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 18]
```diff
@@ -349,6 +337,9 @@ TArray<Worker_InterestOverride> USpatialSender::CreateComponentInterest(AActor*
 		FillComponentInterests(SubobjectInfo, bIsNetOwned, ComponentInterest);
 	}
 
+	ComponentInterest.Add({ SpatialConstants::CLIENT_RPC_ENDPOINT_COMPONENT_ID, bIsNetOwned });
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Did we not need to do this previously because there were only commands in the component?

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialSender.h [hunk 19]
```diff
@@ -87,7 +88,8 @@ class SPATIALGDK_API USpatialSender : public UObject
 
 	// RPC Construction
 	Worker_CommandRequest CreateRPCCommandRequest(UObject* TargetObject, UFunction* Function, void* Parameters, Worker_ComponentId ComponentId, Schema_FieldId CommandIndex, Worker_EntityId& OutEntityId, const UObject*& OutUnresolvedObject, int ReliableRPCIndex);
-	Worker_ComponentUpdate CreateMulticastUpdate(UObject* TargetObject, UFunction* Function, void* Parameters, Worker_ComponentId ComponentId, Schema_FieldId EventIndex, Worker_EntityId& OutEntityId, const UObject*& OutUnresolvedObject);
+	Worker_ComponentUpdate CreateUnreliableRPCUpdate(UObject* TargetObject, UFunction* Function, void* Parameters, Worker_ComponentId ComponentId, Schema_FieldId EventIndex, Worker_EntityId& OutEntityId, const UObject*& OutUnresolvedObject);
+	void WriteRpcPayload(Schema_Object* Object, uint32 Offset, Schema_FieldId Index, FSpatialNetBitWriter PayloadWriter);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialSender.h]:
```suggestion
	void WriteRpcPayload(Schema_Object* Object, uint32 Offset, Schema_FieldId Index, FSpatialNetBitWriter& PayloadWriter);
```

File: SpatialGDK/Extras/schema/rpc_components.schema (hunk 16 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Extras/schema/rpc_components.schema]: It makes me sad too

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp [hunk 20]
```diff
@@ -40,46 +40,63 @@ FORCEINLINE UClass* ResolveClass(FString& ClassPath)
 	return Class;
 }
 
-void USpatialClassInfoManager::CreateClassInfoForClass(UClass* Class)
+ESchemaComponentType GetRpcType(UFunction* RemoteFunction)
```
[the remark this answers is not in the document]
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 21]
```diff
@@ -838,19 +848,39 @@ void USpatialReceiver::OnComponentUpdate(Worker_ComponentUpdateOp& Op)
...
+		Schema_Object* EventData = Schema_IndexObject(EventsObject, SpatialConstants::UNREAL_RPC_ENDPOINT_EVENT_ID, i);
+
+		uint32 Offset = Schema_GetUint32(EventData, SpatialConstants::UNREAL_RPC_PAYLOAD_OFFSET_ID);
+		uint32 Index = Schema_GetUint32(EventData, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_INDEX_ID);
+		TArray<uint8> PayloadData = GetBytesFromSchema(EventData, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_PAYLOAD_ID);
+		int64 CountBits = PayloadData.Num() * 8;
+
+		FUnrealObjectRef ObjectRef(EntityId, Offset);
+
+		UObject* TargetObject = PackageMap->GetObjectFromUnrealObjectRef(ObjectRef).Get();
```
[the remark this answers is not in the document]
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 22]
```diff
@@ -879,14 +909,10 @@ void USpatialReceiver::OnCommandRequest(Worker_CommandRequestOp& Op)
...
+
 	uint32 Offset = 0;
-	bool bFoundOffset = ClassInfoManager->GetOffsetByComponentId(Op.request.component_id, Offset);
-	if (!bFoundOffset)
-	{
-		UE_LOG(LogSpatialReceiver, Warning, TEXT("No offset found for ComponentId %d"), Op.request.component_id);
-		Sender->SendCommandResponse(Op.request_id, Response);
-		return;
-	}
+	Offset = Schema_GetUint32(RequestObject, 1);
```
[the remark this answers is not in the document]
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 23]
```diff
@@ -898,14 +924,9 @@ void USpatialReceiver::OnCommandRequest(Worker_CommandRequestOp& Op)
 
 	const FClassInfo& Info = ClassInfoManager->GetOrCreateClassInfoByObject(TargetObject);
 
-	ESchemaComponentType RPCType = ClassInfoManager->GetCategoryByComponentId(Op.request.component_id);
-	check(RPCType >= SCHEMA_FirstRPC && RPCType <= SCHEMA_LastRPC);
-
-	const TArray<UFunction*>* RPCArray = Info.RPCs.Find(RPCType);
-	check(RPCArray);
-	check((int)CommandIndex - 1 < RPCArray->Num());
+	uint32 Index = Schema_GetUint32(RequestObject, 2);
```
[the remark this answers is not in the document]
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 17 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 18 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Yeah, and the ACLs are reversed with the event-based RPCs. The *sender* has write permission, compared to commands where the *receiver* has write permission.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialSender.h (hunk 19 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialSender.h]: Done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 24]
```diff
@@ -838,19 +848,45 @@ void USpatialReceiver::OnComponentUpdate(Worker_ComponentUpdateOp& Op)
...
+		TArray<uint8> PayloadData = GetBytesFromSchema(EventData, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_PAYLOAD_ID);
+		int64 CountBits = PayloadData.Num() * 8;
+
+		FUnrealObjectRef ObjectRef(EntityId, Offset);
+
+		UObject* TargetObject = PackageMap->GetObjectFromUnrealObjectRef(ObjectRef).Get();
+
+		if (!TargetObject)
+		{
+			UE_LOG(LogSpatialReceiver, Warning, TEXT("HandleUnreliableRPC: Could not find target object: %s, skipping rpc at index: %d"), *ObjectRef.ToString(), Index);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Can we output the function name instead of the index?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Ah no we can't - don't mind me

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h [hunk 25]
```diff
@@ -23,29 +23,30 @@ enum ESchemaComponentType : int32
 	SCHEMA_OwnerOnly,
 	SCHEMA_Handover,
 
+	SCHEMA_Count,
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h]: Should this value be updated to indicate it's only including data components? Where is it even being used?
